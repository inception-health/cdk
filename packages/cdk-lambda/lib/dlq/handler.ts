import {
  LambdaClient,
  InvokeCommandInput,
  InvokeCommand,
} from "@aws-sdk/client-lambda";
import {
  SQSClient,
  ReceiveMessageCommandInput,
  ReceiveMessageCommand,
  DeleteMessageCommandInput,
  DeleteMessageCommand,
} from "@aws-sdk/client-sqs";

const sqs = new SQSClient();
const lambda = new LambdaClient();

export const handler = async () => {
  // Create an SQS client
  const DLQ_URL = process.env.DLQ_URL;
  const LAMBDA_DESTINATION_ARN = process.env.LAMBDA_DESTINATION_ARN;

  if (!DLQ_URL) throw new Error("Missing DLQ_URL environment variable");
  if (!LAMBDA_DESTINATION_ARN)
    throw new Error("Missing LAMBDA_DESTINATION_ARN environment variable");

  try {
    // Receive messages from the DLQ queue
    const receiveParams: ReceiveMessageCommandInput = {
      QueueUrl: DLQ_URL,
      MaxNumberOfMessages: 10,
      WaitTimeSeconds: 10, // in seconds
    };
    const response = await sqs.send(new ReceiveMessageCommand(receiveParams), {
      requestTimeout: 30000, // in milliseconds, 30 seconds
    });

    // Check if there are any messages in the DLQ
    if (response.Messages && response.Messages.length > 0) {
      // Process each message
      for (const message of response.Messages) {
        if (!message.Body) continue;

        // Extract the message body
        const body = JSON.parse(message.Body);

        // Push the message to another Lambda function
        const lambdaParams: InvokeCommandInput = {
          FunctionName: LAMBDA_DESTINATION_ARN,
          InvocationType: "RequestResponse",
          Payload: JSON.stringify(body),
        };
        await lambda.send(new InvokeCommand(lambdaParams));

        // Delete the message from the DLQ
        const deleteParams: DeleteMessageCommandInput = {
          QueueUrl: DLQ_URL,
          ReceiptHandle: message.ReceiptHandle,
        };
        // TODO: This should be replaced with a batch delete
        await sqs.send(new DeleteMessageCommand(deleteParams));
      }
    }
  } catch (error) {
    // Log the error
  }
};
