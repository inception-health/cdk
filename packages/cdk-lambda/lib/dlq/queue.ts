import path from "path";

import { Duration, aws_cloudwatch, aws_kms, aws_sqs } from "aws-cdk-lib";
import { Construct } from "constructs";

import { Lambda } from "../lambda";

export interface DeadLetterQueueProps {
  /**
   * CloudWatch alarm to trigger when the queue has new events.
   *
   * @defaultValue false
   */
  enableAlarm?: boolean;

  /**
   * The encryption key to use for encrypting messages in the queue.
   */
  encryption?: aws_kms.IKey;

  /**
   * Name of the dead letter queue.
   */
  name?: string;

  /**
   * The source lambda function the dead letter queue is attached to.
   *
   * @remarks
   * The source lambda function is the lambda function that will send messages
   * to the dead letter queue.
   */
  source: Lambda;
}

const defaultProps: Partial<DeadLetterQueueProps> = {
  enableAlarm: true,
};

/**
 * A dead letter queue (DLQ) is a queue that receives messages from another
 * queue that the other queue failed to process successfully.
 *
 * In the event that something does go wrong, you need to have a plan to
 * retrieve your messages. So your strategy may change depending on what your
 * event source is. But a common one that works for almost all cases, write a
 * separate lambda function (discussed below) that can pull your DLQ and
 * re-inject those messages back into your original lambda.
 */
export class DeadLetterQueue extends Construct {
  public readonly cdk: {
    alarm: aws_cloudwatch.Alarm | undefined;
    queue: aws_sqs.IQueue;
    queueProcessor: Lambda;
  };

  private readonly props: DeadLetterQueueProps;

  constructor(scope: Construct, id: string, props: DeadLetterQueueProps) {
    super(scope, id);

    this.props = Object.assign({}, defaultProps, props);
    this.cdk = {} as any;

    // Create the redrive function
    this.setupRedriveFunction();

    // Create the dead letter queue
    this.setupDlq();

    // Create the alarm
    this.setupAlarm();

    // Silence the CDK-Nag for the dead letter queue lambda
    // this.silenceCdkNag();
  }

  // !Setup
  private setupRedriveFunction() {
    this.cdk.queueProcessor = new Lambda(this, "redrive", {
      file: path.join(__dirname, "./handler.ts"),
      handler: "handler",
      name: `${this.props.name}--dlq-processor`,
      description: `Redrive function for ${this.props.name} DLQ attached to ${this.props.source.cdk.function.functionName} lambda`,
      environment: {
        LAMBDA_DESTINATION_ARN: this.props.source.cdk.function.functionArn,
      },
    });

    // Grant the lambda function permissions to invoke the target function
    this.props.source.grantInvoke(this.cdk.queueProcessor);
  }

  private setupDlq() {
    this.cdk.queue = new aws_sqs.Queue(this, "queue", {
      queueName: this.props.name,
      retentionPeriod: Duration.days(14),
      encryption: this.props.encryption
        ? aws_sqs.QueueEncryption.KMS
        : aws_sqs.QueueEncryption.KMS_MANAGED,
      encryptionMasterKey: this.props.encryption,
      enforceSSL: true,
      visibilityTimeout: Duration.seconds(30),
    });

    // === Configuration of the Queue Processor ===
    // Grant the lambda function permissions to read from the queue
    this.cdk.queue.grantConsumeMessages(this.cdk.queueProcessor);
    // Add the DLQ URL to the lambda function environment variables
    this.cdk.queueProcessor.addEnvironment("DLQ_URL", this.cdk.queue.queueUrl);
  }

  private setupAlarm() {
    if (!this.props.enableAlarm) return;
    this.cdk.alarm = new aws_cloudwatch.Alarm(this, "alarm", {
      alarmName: `${this.props.name}--dlq-alarm`,
      alarmDescription: `Alarm for ${this.props.name} dead letter queue`,
      metric: this.cdk.queue.metricApproximateNumberOfMessagesVisible(),
      threshold: 5,
      evaluationPeriods: 1,
      treatMissingData: aws_cloudwatch.TreatMissingData.NOT_BREACHING,
    });
  }

  private silenceCdkNag() {
    // this.cdk.queue.node.applyAspect({
    //   visit(node: Construct) {
    //     if (node.node.metadata.length) {
    //       node.node.metadata.forEach((meta) => {
    //         if (meta.type === "cdk-nag") {
    //           node.node.metadata.splice(node.node.metadata.indexOf(meta), 1);
    //         }
    //       });
    //     }
    //   },
    // });
  }
}
