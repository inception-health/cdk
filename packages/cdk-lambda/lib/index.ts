export { Lambda } from "./lambda";
export type { LambdaProps } from "./lambda";

export { DeadLetterQueue } from "./dlq/queue";
export type { DeadLetterQueueProps } from "./dlq/queue";

export { StandardNetwork } from "./network/network";
export type { StandardNetworkProps } from "./network/network";
