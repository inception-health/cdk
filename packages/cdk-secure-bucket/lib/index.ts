export * from "./s3.secure";
export * from "./lifecycle";
