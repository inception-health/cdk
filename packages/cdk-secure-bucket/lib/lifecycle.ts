import { aws_s3, Duration } from "aws-cdk-lib";

// The lifecycle policy for items should be managed cohesively
export class LifecycleMode {
  static readonly NONE = new LifecycleMode();

  static readonly BASIC = new LifecycleMode({
    enabled: true,
    transitions: [
      {
        storageClass: aws_s3.StorageClass.INFREQUENT_ACCESS,
        transitionAfter: Duration.days(360),
      },
      {
        storageClass: aws_s3.StorageClass.GLACIER,
        transitionAfter: Duration.days(360 * 3),
      },
    ],
  });

  static readonly INTELLIGENT = new LifecycleMode({
    enabled: true,
    transitions: [
      {
        storageClass: aws_s3.StorageClass.INTELLIGENT_TIERING,
        transitionAfter: Duration.days(360),
      },
      {
        storageClass: aws_s3.StorageClass.GLACIER,
        transitionAfter: Duration.days(360 * 3),
      },
    ],
  });

  private constructor(public readonly rule?: aws_s3.LifecycleRule) {}
}
