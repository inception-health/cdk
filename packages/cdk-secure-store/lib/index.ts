export type { SecureStoreProps } from "./store";
export { SecureStore } from "./store";
