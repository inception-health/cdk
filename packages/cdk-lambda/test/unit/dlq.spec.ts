import path from "path";

import { CfnElement, Stack, aws_kms, Aspects, App } from "aws-cdk-lib";
import { Match, Template, Annotations } from "aws-cdk-lib/assertions";
import { HIPAASecurityChecks } from "cdk-nag";

import { DeadLetterQueue } from "../../lib/dlq/queue";
import { Lambda } from "../../lib/lambda";

describe("Dead Letter Queue", () => {
  describe("General", () => {
    it("should match the snapshot", () => {
      // Given
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });

      // When
      new DeadLetterQueue(stack, "dlq", {
        name: "some-lambda-function",
        source: source,
      });

      // Then
      expect(Template.fromStack(stack)).toMatchSnapshot();
    });

    it("should setup an alarm", () => {
      // Given
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });

      // When
      new DeadLetterQueue(stack, "dlq", {
        name: "some-lambda-function",
        source: source,
      });

      // Then
      const template = Template.fromStack(stack);
      template.resourceCountIs("AWS::CloudWatch::Alarm", 1);
    });
  });

  describe("Encryption", () => {
    it("should use SSE-KMS if no encryption key is provided", () => {
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });
      const dlqName = "some-dlq-name";

      // When
      new DeadLetterQueue(stack, "dlq", {
        name: dlqName,
        source: source,
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::SQS::Queue", {
        KmsMasterKeyId: "alias/aws/sqs",
      });
    });

    it("should use custom KMS", () => {
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });
      const dlqName = "some-dlq-name";
      const key = new aws_kms.Key(stack, "key", {
        alias: "my-key",
      });

      // When
      new DeadLetterQueue(stack, "dlq", {
        name: dlqName,
        encryption: key,
        source: source,
      });

      // Then
      const template = Template.fromStack(stack);

      template.hasResourceProperties("AWS::SQS::Queue", {
        KmsMasterKeyId: {
          "Fn::GetAtt": [Match.anyValue(), "Arn"],
        },
      });
    });
  });

  describe("DLQ Lambda Handler", () => {
    it("should have a lambda function", () => {
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });

      // When
      new DeadLetterQueue(stack, "dlq", {
        name: "some-dlq-name",
        source: source,
      });

      // Then
      const template = Template.fromStack(stack);
      template.resourceCountIs("AWS::Lambda::Function", 2);
    });

    it("should have appropriate environment variables", () => {
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });

      // When
      const dlq = new DeadLetterQueue(stack, "dlq", {
        name: "some-dlq-name",
        source: source,
      });

      // Expect
      const expectedDqlRef = stack.getLogicalId(
        dlq.cdk.queue.node.defaultChild as CfnElement,
      );
      const expectedLambdaRef = stack.getLogicalId(
        source.cdk.function.node.defaultChild as CfnElement,
      );

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Lambda::Function", {
        Environment: {
          Variables: {
            DLQ_URL: {
              Ref: expectedDqlRef,
            },
            LAMBDA_DESTINATION_ARN: {
              "Fn::GetAtt": [expectedLambdaRef, "Arn"],
            },
          },
        },
      });
    });

    it("should have invocation permission to the source function", () => {
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });

      // When
      const dlq = new DeadLetterQueue(stack, "dlq", {
        name: "some-dlq-name",
        source: source,
      });

      // Expect
      const expectedLambdaRef = stack.getLogicalId(
        source.cdk.function.node.defaultChild as CfnElement,
      );
      const expectedDlqHandlerRoleRef = stack.getLogicalId(
        dlq.cdk.queueProcessor.cdk.function.role?.node
          .defaultChild as CfnElement,
      );

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::IAM::Policy", {
        PolicyDocument: {
          Statement: Match.arrayWith([
            {
              Action: "lambda:InvokeFunction",
              Effect: "Allow",
              Resource: [
                {
                  "Fn::GetAtt": [expectedLambdaRef, "Arn"],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [expectedLambdaRef, "Arn"],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ]),
          Version: "2012-10-17",
        },
        Roles: [
          {
            Ref: expectedDlqHandlerRoleRef,
          },
        ],
      });
    });

    it("should have permissions to read from the DLQ queue", () => {
      const stack = new Stack();
      const source = new Lambda(stack, "handler", {
        file: path.join(__dirname, "./mocks/function.ts"),
        name: "some-lambda-function",
      });

      // When
      const dlq = new DeadLetterQueue(stack, "dlq", {
        name: "some-dlq-name",
        source: source,
      });

      // Expect
      const expectedDqlRef = stack.getLogicalId(
        dlq.cdk.queue.node.defaultChild as CfnElement,
      );
      const expectedDlqHandlerRoleRef = stack.getLogicalId(
        dlq.cdk.queueProcessor.cdk.function.role?.node
          .defaultChild as CfnElement,
      );
      // Then
      const template = Template.fromStack(stack);
      // 2. Should have a Lambda permission to retrieve messages from the DLQ. and
      //    to send messages to the original Lambda function.
      template.hasResourceProperties("AWS::IAM::Policy", {
        PolicyDocument: {
          Statement: Match.arrayWith([
            {
              Action: [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              Effect: "Allow",
              Resource: {
                "Fn::GetAtt": [expectedDqlRef, "Arn"],
              },
            },
          ]),
          Version: "2012-10-17",
        },
        Roles: [
          {
            Ref: expectedDlqHandlerRoleRef,
          },
        ],
      });
    });
  });
});
