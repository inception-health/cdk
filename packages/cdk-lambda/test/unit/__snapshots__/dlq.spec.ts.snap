// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dead Letter Queue General should match the snapshot 1`] = `
{
  "Mappings": {
    "CloudwatchlambdainsightsversionMap": {
      "af-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:af-south-1:012438385374:layer:LambdaInsightsExtension-Arm64:18",
      },
      "ap-east-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-east-1:519774774795:layer:LambdaInsightsExtension-Arm64:18",
      },
      "ap-northeast-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-northeast-1:580247275435:layer:LambdaInsightsExtension-Arm64:31",
      },
      "ap-northeast-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-northeast-2:580247275435:layer:LambdaInsightsExtension-Arm64:19",
      },
      "ap-northeast-3": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-northeast-3:194566237122:layer:LambdaInsightsExtension-Arm64:17",
      },
      "ap-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-south-1:580247275435:layer:LambdaInsightsExtension-Arm64:22",
      },
      "ap-south-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-south-2:891564319516:layer:LambdaInsightsExtension-Arm64:6",
      },
      "ap-southeast-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-southeast-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "ap-southeast-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-southeast-2:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "ap-southeast-3": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-southeast-3:439286490199:layer:LambdaInsightsExtension-Arm64:18",
      },
      "ca-central-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ca-central-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "eu-central-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-central-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "eu-north-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-north-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "eu-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-south-1:339249233099:layer:LambdaInsightsExtension-Arm64:18",
      },
      "eu-south-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-south-2:352183217350:layer:LambdaInsightsExtension-Arm64:6",
      },
      "eu-west-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-west-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "eu-west-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-west-2:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "eu-west-3": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-west-3:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "me-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:me-south-1:285320876703:layer:LambdaInsightsExtension-Arm64:18",
      },
      "sa-east-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:sa-east-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "us-east-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-east-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "us-east-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-east-2:580247275435:layer:LambdaInsightsExtension-Arm64:22",
      },
      "us-gov-east-1": {
        "1x0x333x0xarm64": "arn:aws-us-gov:lambda:us-gov-east-1:122132214140:layer:LambdaInsightsExtension-Arm64:3",
      },
      "us-gov-west-1": {
        "1x0x333x0xarm64": "arn:aws-us-gov:lambda:us-gov-west-1:751350123760:layer:LambdaInsightsExtension-Arm64:3",
      },
      "us-west-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-west-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "us-west-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-west-2:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "dlqalarmEF73E847": {
      "Properties": {
        "AlarmDescription": "Alarm for some-lambda-function dead letter queue",
        "AlarmName": "some-lambda-function--dlq-alarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "dlqqueueBDF0D8BE",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 5,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "dlqqueueBDF0D8BE": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": "alias/aws/sqs",
        "MessageRetentionPeriod": 1209600,
        "QueueName": "some-lambda-function",
        "VisibilityTimeout": 30,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "dlqqueuePolicy2C452EE9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "dlqqueueBDF0D8BE",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "dlqqueueBDF0D8BE",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "dlqredrivefunction2BC30FF7": {
      "DependsOn": [
        "dlqredriveroleDefaultPolicyB3D3172E",
        "dlqredriverole73EF79BD",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "381b9a88b771f3cfd8a6bbfffcc8fb8bda0d6c48b770c58919bca944ad838115.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Redrive function for some-lambda-function DLQ attached to ",
              {
                "Ref": "handlerfunctionC9E083FB",
              },
              " lambda",
            ],
          ],
        },
        "Environment": {
          "Variables": {
            "DLQ_URL": {
              "Ref": "dlqqueueBDF0D8BE",
            },
            "LAMBDA_DESTINATION_ARN": {
              "Fn::GetAtt": [
                "handlerfunctionC9E083FB",
                "Arn",
              ],
            },
          },
        },
        "EphemeralStorage": {
          "Size": 512,
        },
        "FunctionName": "some-lambda-function--dlq-processor",
        "Handler": "index.handler",
        "Layers": [
          {
            "Fn::FindInMap": [
              "CloudwatchlambdainsightsversionMap",
              {
                "Ref": "AWS::Region",
              },
              "1x0x333x0xarm64",
            ],
          },
        ],
        "LoggingConfig": {
          "ApplicationLogLevel": "INFO",
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "dlqredriveloggroup57A2A78B",
          },
          "SystemLogLevel": "INFO",
        },
        "MemorySize": 128,
        "ReservedConcurrentExecutions": 20,
        "Role": {
          "Fn::GetAtt": [
            "dlqredriverole73EF79BD",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "dlqredriveloggroup57A2A78B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "/aws/lambda/some-lambda-function--dlq-processor",
        "RetentionInDays": 60,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "dlqredriverole73EF79BD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLambdaInsightsExecutionRolePolicy",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "dlqredriveroleDefaultPolicyB3D3172E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "handlerfunctionC9E083FB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "handlerfunctionC9E083FB",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "dlqqueueBDF0D8BE",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "dlqredriveroleDefaultPolicyB3D3172E",
        "Roles": [
          {
            "Ref": "dlqredriverole73EF79BD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "handlerfunctionC9E083FB": {
      "DependsOn": [
        "handlerroleDefaultPolicy04A931DB",
        "handlerroleB68B03AE",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7671fef3231da5b8460e43bcffeb2c9d9d7763730b7734f98ef6960dfeda58d3.zip",
        },
        "EphemeralStorage": {
          "Size": 512,
        },
        "FunctionName": "some-lambda-function",
        "Handler": "index.handler",
        "Layers": [
          {
            "Fn::FindInMap": [
              "CloudwatchlambdainsightsversionMap",
              {
                "Ref": "AWS::Region",
              },
              "1x0x333x0xarm64",
            ],
          },
        ],
        "LoggingConfig": {
          "ApplicationLogLevel": "INFO",
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "handlerloggroup94E54AD1",
          },
          "SystemLogLevel": "INFO",
        },
        "MemorySize": 128,
        "ReservedConcurrentExecutions": 20,
        "Role": {
          "Fn::GetAtt": [
            "handlerroleB68B03AE",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "handlerloggroup94E54AD1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "/aws/lambda/some-lambda-function",
        "RetentionInDays": 60,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "handlerroleB68B03AE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLambdaInsightsExecutionRolePolicy",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "handlerroleDefaultPolicy04A931DB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "handlerroleDefaultPolicy04A931DB",
        "Roles": [
          {
            "Ref": "handlerroleB68B03AE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
